<img src = "https://global-uploads.webflow.com/5f0d53c042a9ed6288de7f8d/5f6337ae2cfaa10946ceeb06_Hacktiv8%20logo%20horizontal%2001%20black-p-500.png" width = 400>

# Assignment 2

by Ahmat Zainul Muttaqin

PYTN-KS019

Studi Independen - Kampus Merdeka

## Project Overview

Using what you’ve learned; download the NYC Property Sales Dataset from Kaggle. This dataset is a record of every building or building unit (apartment, etc.) sold in the New York City property market over a 12-month period.

This dataset contains the location, address, type, sale price, and sale date of building units sold. A reference on the trickier fields:

- BOROUGH: A digit code for the borough the property is located in; in order these are Manhattan (1), Bronx (2), Brooklyn (3), Queens (4), and Staten Island (5).
- BLOCK; LOT: The combination of borough, block, and lot forms a unique key for property in New York City. Commonly called a BBL.
- BUILDING CLASS AT PRESENT and BUILDING CLASS AT TIME OF SALE: The type of building at various points in time.

Note that because this is a financial transaction dataset, there are some points that need to be kept in mind:

- Many sales occur with a nonsensically small dollar amount: $0 most commonly. These sales are actually transfers of deeds between parties: for example, parents transferring ownership to their home to a child after moving out for retirement.
- This dataset uses the financial definition of a building/building unit, for tax purposes. In case a single entity owns the building in question, a sale covers the value of the entire building. In case a building is owned piecemeal by its residents (a condominium), a sale refers to a single apartment (or group of apartments) owned by some individual.

Formulate a question and derive a statistical hypothesis test to answer the question. You have to demonstrate that you’re able to make decisions using data in a scientific manner. Examples of questions can be:

- Is there a difference in unit sold between property built in 1900-2000 and 2001 so on?
- Is there a difference in unit sold based on building category?
- What can you discover about New York City real estate by looking at a year's worth of raw transaction records? Can you spot trends in the market?

## Data Preparation

Before we start, we have to import the libraries that we going to use.

In [3]:
import numpy as np
import pandas as pd

#---------- For Descriptive Statistics #----------
import math
import statistics
import scipy.stats
import matplotlib as mpl
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm

print("Library sucessfully imported!")

Library sucessfully imported!


Dataset is about NYC Property Sales, [Download Here](https://www.kaggle.com/new-york-city/nyc-property-sales)

Dataset file must in the same folder with this file to run this code for importing datasets using Pandas.

In [4]:
df = pd.read_csv('nyc-rolling-sales.csv')

print('Dataset is ready!')

Dataset is ready!


View the top 5 rows of dataset

In [6]:
df.head()

Unnamed: 0  BOROUGH   NEIGHBORHOOD  \
0           4        1  ALPHABET CITY   
1           5        1  ALPHABET CITY   
2           6        1  ALPHABET CITY   
3           7        1  ALPHABET CITY   
4           8        1  ALPHABET CITY   

                       BUILDING CLASS CATEGORY TAX CLASS AT PRESENT  BLOCK  \
0  07 RENTALS - WALKUP APARTMENTS                                2A    392   
1  07 RENTALS - WALKUP APARTMENTS                                 2    399   
2  07 RENTALS - WALKUP APARTMENTS                                 2    399   
3  07 RENTALS - WALKUP APARTMENTS                                2B    402   
4  07 RENTALS - WALKUP APARTMENTS                                2A    404   

   LOT EASE-MENT BUILDING CLASS AT PRESENT                 ADDRESS  ...  \
0    6                                  C2            153 AVENUE B  ...   
1   26                                  C7   234 EAST 4TH   STREET  ...   
2   39                                  C7   197 EAST 3RD   STREET  ...   
3   21                                  C4     154 EAST 7TH STREET  ...   
4   55                                  C2  301 EAST 10TH   STREET  ...   

  RESIDENTIAL UNITS  COMMERCIAL UNITS  TOTAL UNITS  LAND SQUARE FEET  \
0                 5                 0            5              1633   
1                28                 3           31              4616   
2                16                 1           17              2212   
3                10                 0           10              2272   
4                 6                 0            6              2369   

   GROSS SQUARE FEET YEAR BUILT TAX CLASS AT TIME OF SALE  \
0               6440       1900                         2   
1              18690       1900                         2   
2               7803       1900                         2   
3               6794       1913                         2   
4               4615       1900                         2   

   BUILDING CLASS AT TIME OF SALE  SALE PRICE            SALE DATE  
0                              C2     6625000  2017-07-19 00:00:00  
1                              C7         -    2016-12-14 00:00:00  
2                              C7         -    2016-12-09 00:00:00  
3                              C4     3936272  2016-09-23 00:00:00  
4                              C2     8000000  2016-11-17 00:00:00  

[5 rows x 22 columns]

View the bottom 5 rows of dataset

In [7]:
df.tail()

Unnamed: 0  BOROUGH NEIGHBORHOOD  \
84543        8409        5      WOODROW   
84544        8410        5      WOODROW   
84545        8411        5      WOODROW   
84546        8412        5      WOODROW   
84547        8413        5      WOODROW   

                           BUILDING CLASS CATEGORY TAX CLASS AT PRESENT  \
84543  02 TWO FAMILY DWELLINGS                                        1   
84544  02 TWO FAMILY DWELLINGS                                        1   
84545  02 TWO FAMILY DWELLINGS                                        1   
84546  22 STORE BUILDINGS                                             4   
84547  35 INDOOR PUBLIC AND CULTURAL FACILITIES                       4   

       BLOCK  LOT EASE-MENT BUILDING CLASS AT PRESENT                ADDRESS  \
84543   7349   34                                  B9          37 QUAIL LANE   
84544   7349   78                                  B9       32 PHEASANT LANE   
84545   7351   60                                  B2       49 PITNEY AVENUE   
84546   7100   28                                  K6  2730 ARTHUR KILL ROAD   
84547   7105  679                                  P9      155 CLAY PIT ROAD   

       ... RESIDENTIAL UNITS  COMMERCIAL UNITS  TOTAL UNITS  LAND SQUARE FEET  \
84543  ...                 2                 0            2              2400   
84544  ...                 2                 0            2              2498   
84545  ...                 2                 0            2              4000   
84546  ...                 0                 7            7            208033   
84547  ...                 0                 1            1             10796   

       GROSS SQUARE FEET YEAR BUILT TAX CLASS AT TIME OF SALE  \
84543               2575       1998                         1   
84544               2377       1998                         1   
84545               1496       1925                         1   
84546              64117       2001                         4   
84547               2400       2006                         4   

       BUILDING CLASS AT TIME OF SALE  SALE PRICE            SALE DATE  
84543                              B9      450000  2016-11-28 00:00:00  
84544                              B9      550000  2017-04-21 00:00:00  
84545                              B2      460000  2017-07-05 00:00:00  
84546                              K6    11693337  2016-12-21 00:00:00  
84547                              P9       69300  2016-10-27 00:00:00  

[5 rows x 22 columns]

View all columns and data type

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 84548 entries, 0 to 84547
Data columns (total 22 columns):
 #   Column                          Non-Null Count  Dtype 
---  ------                          --------------  ----- 
 0   Unnamed: 0                      84548 non-null  int64 
 1   BOROUGH                         84548 non-null  int64 
 2   NEIGHBORHOOD                    84548 non-null  object
 3   BUILDING CLASS CATEGORY         84548 non-null  object
 4   TAX CLASS AT PRESENT            84548 non-null  object
 5   BLOCK                           84548 non-null  int64 
 6   LOT                             84548 non-null  int64 
 7   EASE-MENT                       84548 non-null  object
 8   BUILDING CLASS AT PRESENT       84548 non-null  object
 9   ADDRESS                         84548 non-null  object
 10  APARTMENT NUMBER                84548 non-null  object
 11  ZIP CODE                        84548 non-null  int64 
 12  RESIDENTIAL UNITS               84548 non-null

In [4]:
df.shape

(84548, 22)

View header list

In [8]:
df.columns.values

array(['Unnamed: 0', 'BOROUGH', 'NEIGHBORHOOD', 'BUILDING CLASS CATEGORY',
       'TAX CLASS AT PRESENT', 'BLOCK', 'LOT', 'EASE-MENT',
       'BUILDING CLASS AT PRESENT', 'ADDRESS', 'APARTMENT NUMBER',
       'ZIP CODE', 'RESIDENTIAL UNITS', 'COMMERCIAL UNITS', 'TOTAL UNITS',
       'LAND SQUARE FEET', 'GROSS SQUARE FEET', 'YEAR BUILT',
       'TAX CLASS AT TIME OF SALE', 'BUILDING CLASS AT TIME OF SALE',
       'SALE PRICE', 'SALE DATE'], dtype=object)

## Pre-Processing

Next, remove the EASE-MENT and Unnamed: 0 columns because this data is not required and there are no values within those columns. By removing these two columns from the original table, we can obtain accurate results.

In [9]:
# Remove the EASE-MENT and Unnamed: 0 columns as they are not very necessary
df_no_east = df.drop(["Unnamed: 0", "EASE-MENT"], axis=1)

# Check the Data
df_no_east.head(5)

BOROUGH   NEIGHBORHOOD                      BUILDING CLASS CATEGORY  \
0        1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS                
1        1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS                
2        1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS                
3        1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS                
4        1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS                

  TAX CLASS AT PRESENT  BLOCK  LOT BUILDING CLASS AT PRESENT  \
0                   2A    392    6                        C2   
1                    2    399   26                        C7   
2                    2    399   39                        C7   
3                   2B    402   21                        C4   
4                   2A    404   55                        C2   

                  ADDRESS APARTMENT NUMBER  ZIP CODE  RESIDENTIAL UNITS  \
0            153 AVENUE B                      10009                  5   
1   234 EAST 4TH   STREET                      10009                 28   
2   197 EAST 3RD   STREET                      10009                 16   
3     154 EAST 7TH STREET                      10009                 10   
4  301 EAST 10TH   STREET                      10009                  6   

   COMMERCIAL UNITS  TOTAL UNITS LAND SQUARE FEET GROSS SQUARE FEET  \
0                 0            5             1633              6440   
1                 3           31             4616             18690   
2                 1           17             2212              7803   
3                 0           10             2272              6794   
4                 0            6             2369              4615   

   YEAR BUILT  TAX CLASS AT TIME OF SALE BUILDING CLASS AT TIME OF SALE  \
0        1900                          2                             C2   
1        1900                          2                             C7   
2        1900                          2                             C7   
3        1913                          2                             C4   
4        1900                          2                             C2   

  SALE PRICE            SALE DATE  
0    6625000  2017-07-19 00:00:00  
1        -    2016-12-14 00:00:00  
2        -    2016-12-09 00:00:00  
3    3936272  2016-09-23 00:00:00  
4    8000000  2016-11-17 00:00:00

Viewing information about the data in the DataFrame.

In [6]:
df_no_east.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 84548 entries, 0 to 84547
Data columns (total 20 columns):
 #   Column                          Non-Null Count  Dtype 
---  ------                          --------------  ----- 
 0   BOROUGH                         84548 non-null  int64 
 1   NEIGHBORHOOD                    84548 non-null  object
 2   BUILDING CLASS CATEGORY         84548 non-null  object
 3   TAX CLASS AT PRESENT            84548 non-null  object
 4   BLOCK                           84548 non-null  int64 
 5   LOT                             84548 non-null  int64 
 6   BUILDING CLASS AT PRESENT       84548 non-null  object
 7   ADDRESS                         84548 non-null  object
 8   APARTMENT NUMBER                84548 non-null  object
 9   ZIP CODE                        84548 non-null  int64 
 10  RESIDENTIAL UNITS               84548 non-null  int64 
 11  COMMERCIAL UNITS                84548 non-null  int64 
 12  TOTAL UNITS                     84548 non-null

Checking the data type of each column to facilitate grouping columns with the same data type, which aims to simplify the data processing.

In [10]:
# Mengelompokan data dengan type data object
data_cat_obj = [
    "NEIGHBORHOOD", 
    "BUILDING CLASS CATEGORY", 
    "TAX CLASS AT PRESENT",
    "BLOCK",
    "LOT",
    "BUILDING CLASS AT PRESENT",
    "ADDRESS",
    "APARTMENT NUMBER",
    "BUILDING CLASS AT TIME OF SALE",
    "SALE DATE"
]
for data_group in data_cat_obj:
    df_no_east[data_group] = df_no_east[data_group].astype('category')

# Mengelompokan data dengan type data integer dan mengubahnya ke numeric
data_cat_int = [
    "BOROUGH",
    "SALE PRICE",
    "LAND SQUARE FEET",
    "GROSS SQUARE FEET",
    
]
for data_group in data_cat_int:
    df_no_east[data_group] = pd.to_numeric(df_no_east[data_group], errors='coerce')

# Mengubah type data SALE DATE dari Object menjadi Date
df_no_east["SALE DATE"] = pd.to_datetime(df_no_east["SALE DATE"])

<b>Note</b>

To avoid errors that may occur in the "SALE PRICE" column, especially when it contains the value "(-)," you can handle it by adding the parameter "errors='coerce'." With this parameter, any invalid values will be converted to NaN (Not a Number).

<b>More Information</b>

Parameters:
- <b>arg:</b> scalar, list, tuple, 1-d array, or Series
  - Argument to be converted.

- <b>errors:</b> {'ignore', 'raise', 'coerce'}, default 'raise'
  - If 'raise,' invalid parsing will raise an exception.
  - If 'coerce,' invalid parsing will be set as NaN.
  - If 'ignore,' invalid parsing will return the input.

Next, let's check the number of data entries that have null (missing) values.

In [11]:
df_no_east.isnull().sum()

BOROUGH                               0
NEIGHBORHOOD                          0
BUILDING CLASS CATEGORY               0
TAX CLASS AT PRESENT                  0
BLOCK                                 0
LOT                                   0
BUILDING CLASS AT PRESENT             0
ADDRESS                               0
APARTMENT NUMBER                      0
ZIP CODE                              0
RESIDENTIAL UNITS                     0
COMMERCIAL UNITS                      0
TOTAL UNITS                           0
LAND SQUARE FEET                  26252
GROSS SQUARE FEET                 27612
YEAR BUILT                            0
TAX CLASS AT TIME OF SALE             0
BUILDING CLASS AT TIME OF SALE        0
SALE PRICE                        14561
SALE DATE                             0
dtype: int64

From the check, we found null values in the "LAND SQUARE FEET" column, with 26,252 missing values, the "GROSS SQUARE FEET" column with 27,612 missing values, and the "SALE PRICE" column with 14,561 missing values.

With this many missing values, we can replace the null values with the mean (average) value of each respective column to ensure the accuracy of the resulting data.

In [13]:
df_no_east['LAND SQUARE FEET'].fillna(df_no_east['LAND SQUARE FEET'].mean(), inplace=True)
df_no_east['GROSS SQUARE FEET'].fillna(df_no_east['GROSS SQUARE FEET'].mean(), inplace=True)
df_no_east['SALE PRICE'].fillna(df_no_east['SALE PRICE'].mean(), inplace=True)

We are counting the number of null values in the columns 'LAND SQUARE FEET', 'GROSS SQUARE FEET', and 'SALE PRICE' to see how many missing values are in each column.

In [14]:
df_no_east[['LAND SQUARE FEET', 'GROSS SQUARE FEET', 'SALE PRICE']].isnull().sum()

LAND SQUARE FEET     0
GROSS SQUARE FEET    0
SALE PRICE           0
dtype: int64

<b>Explanation<b>
    
By filling in the null values with the mean of each column, we have successfully replaced the missing values in the columns 'LAND SQUARE FEET', 'GROSS SQUARE FEET', and 'SALE PRICE'. As a result, there are now 0 null values in these columns.

Displaying the First 5 Rows

In [15]:
df_no_east.head()

BOROUGH   NEIGHBORHOOD                      BUILDING CLASS CATEGORY  \
0        1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS                
1        1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS                
2        1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS                
3        1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS                
4        1  ALPHABET CITY  07 RENTALS - WALKUP APARTMENTS                

  TAX CLASS AT PRESENT BLOCK LOT BUILDING CLASS AT PRESENT  \
0                   2A   392   6                        C2   
1                    2   399  26                        C7   
2                    2   399  39                        C7   
3                   2B   402  21                        C4   
4                   2A   404  55                        C2   

                  ADDRESS APARTMENT NUMBER  ZIP CODE  RESIDENTIAL UNITS  \
0            153 AVENUE B                      10009                  5   
1   234 EAST 4TH   STREET                      10009                 28   
2   197 EAST 3RD   STREET                      10009                 16   
3     154 EAST 7TH STREET                      10009                 10   
4  301 EAST 10TH   STREET                      10009                  6   

   COMMERCIAL UNITS  TOTAL UNITS  LAND SQUARE FEET  GROSS SQUARE FEET  \
0                 0            5            1633.0             6440.0   
1                 3           31            4616.0            18690.0   
2                 1           17            2212.0             7803.0   
3                 0           10            2272.0             6794.0   
4                 0            6            2369.0             4615.0   

   YEAR BUILT  TAX CLASS AT TIME OF SALE BUILDING CLASS AT TIME OF SALE  \
0        1900                          2                             C2   
1        1900                          2                             C7   
2        1900                          2                             C7   
3        1913                          2                             C4   
4        1900                          2                             C2   

     SALE PRICE  SALE DATE  
0  6.625000e+06 2017-07-19  
1  1.276456e+06 2016-12-14  
2  1.276456e+06 2016-12-09  
3  3.936272e+06 2016-09-23  
4  8.000000e+06 2016-11-17

Count of Residential Units

In [13]:
df_no_east['RESIDENTIAL UNITS'].value_counts()

RESIDENTIAL UNITS
1      34722
0      24783
2      16049
3       4608
4       1346
       ...  
179        1
75         1
193        1
153        1
128        1
Name: count, Length: 176, dtype: int64

Data exploration

In [16]:
df_no_east[df_no_east['COMMERCIAL UNITS']==0].describe()

BOROUGH      ZIP CODE  RESIDENTIAL UNITS  COMMERCIAL UNITS  \
count  79429.000000  79429.000000       79429.000000           79429.0   
mean       3.004797  10721.690503           1.717257               0.0   
std        1.299437   1320.121756          10.146362               0.0   
min        1.000000      0.000000           0.000000               0.0   
25%        2.000000  10304.000000           0.000000               0.0   
50%        3.000000  11209.000000           1.000000               0.0   
75%        4.000000  11357.000000           2.000000               0.0   
max        5.000000  11694.000000         889.000000               0.0   

        TOTAL UNITS  LAND SQUARE FEET  GROSS SQUARE FEET    YEAR BUILT  \
count  79429.000000      7.942900e+04       7.942900e+04  79429.000000   
mean       1.750217      3.423023e+03       2.964567e+03   1780.395787   
std       10.142353      2.394616e+04       9.865621e+03    551.999810   
min        0.000000      0.000000e+00       0.000000e+00      0.000000   
25%        1.000000      2.000000e+03       1.312000e+03   1920.000000   
50%        1.000000      3.263000e+03       2.400000e+03   1940.000000   
75%        2.000000      3.941676e+03       4.045707e+03   1967.000000   
max      889.000000      4.252327e+06       1.378125e+06   2017.000000   

       TAX CLASS AT TIME OF SALE    SALE PRICE  
count               79429.000000  7.942900e+04  
mean                    1.584258  1.040170e+06  
std                     0.720488  3.028813e+06  
min                     1.000000  0.000000e+00  
25%                     1.000000  3.000000e+05  
50%                     1.000000  6.550000e+05  
75%                     2.000000  1.276456e+06  
max                     4.000000  3.450000e+08

Data Exploration for Residential Units Equal to 0

In [17]:
df_no_east[df_no_east['RESIDENTIAL UNITS']==0].describe()

BOROUGH      ZIP CODE  RESIDENTIAL UNITS  COMMERCIAL UNITS  \
count  24783.000000  24783.000000            24783.0      24783.000000   
mean       2.537869  10318.242182                0.0          0.492515   
std        1.335198   2136.962746                0.0         16.073953   
min        1.000000      0.000000                0.0          0.000000   
25%        1.000000  10023.000000                0.0          0.000000   
50%        3.000000  11004.000000                0.0          0.000000   
75%        4.000000  11354.000000                0.0          0.000000   
max        5.000000  11694.000000                0.0       2261.000000   

        TOTAL UNITS  LAND SQUARE FEET  GROSS SQUARE FEET    YEAR BUILT  \
count  24783.000000      2.478300e+04       2.478300e+04  24783.000000   
mean       0.598152      5.465813e+03       5.431326e+03   1674.012145   
std       16.073655      5.098233e+04       3.262132e+04    680.924383   
min        0.000000      0.000000e+00       0.000000e+00      0.000000   
25%        0.000000      2.047000e+03       2.775000e+02   1921.000000   
50%        0.000000      3.941676e+03       4.045707e+03   1950.000000   
75%        0.000000      3.941676e+03       4.045707e+03   1962.000000   
max     2261.000000      4.252327e+06       1.617206e+06   2017.000000   

       TAX CLASS AT TIME OF SALE    SALE PRICE  
count               24783.000000  2.478300e+04  
mean                    2.423879  1.575554e+06  
std                     0.913323  1.807579e+07  
min                     1.000000  0.000000e+00  
25%                     2.000000  2.150000e+05  
50%                     2.000000  5.200000e+05  
75%                     2.000000  1.276456e+06  
max                     4.000000  2.210000e+09

Count the occurrences of each unique value in the "BOROUGH" column

In [22]:
df_no_east['BOROUGH'].value_counts()

BOROUGH
4    26736
3    24047
1    18306
5     8410
2     7049
Name: count, dtype: int64

Group the DataFrame by the "BOROUGH" column and calculate the mean of "SALE PRICE" for each borough

In [20]:
result = df_no_east.groupby('BOROUGH')['SALE PRICE'].mean()
result

BOROUGH
1    2.902477e+06
2    5.901936e+05
3    8.344884e+05
4    9.063757e+05
5    7.525600e+05
Name: SALE PRICE, dtype: float64

Group the DataFrame by the "BOROUGH" column and find the mode (most frequent values) for each group.Then reset the index and drop the previous index to get a DataFrame with the modes for each borough

In [24]:
df_no_east.groupby('BOROUGH').apply(pd.DataFrame.mode).reset_index(drop=True)

BOROUGH             NEIGHBORHOOD  \
0        1  UPPER EAST SIDE (59-79)   
1        2                RIVERDALE   
2        3       BEDFORD STUYVESANT   
3        4           FLUSHING-NORTH   
4        5              GREAT KILLS   

                       BUILDING CLASS CATEGORY TAX CLASS AT PRESENT BLOCK LOT  \
0  13 CONDOS - ELEVATOR APARTMENTS                                2    16   1   
1  02 TWO FAMILY DWELLINGS                                        1  3944   1   
2  02 TWO FAMILY DWELLINGS                                        1  2135   1   
3  01 ONE FAMILY DWELLINGS                                        1  5066   1   
4  01 ONE FAMILY DWELLINGS                                        1  2450   1   

  BUILDING CLASS AT PRESENT                 ADDRESS APARTMENT NUMBER  \
0                        R4  169 WEST 95TH   STREET                    
1                        D4     2287 JOHNSON AVENUE                    
2                        R4         429 KENT AVENUE                    
3                        A1      131-05 40TH   ROAD                    
4                        A5   1235 FOREST HILL ROAD                    

   ZIP CODE  RESIDENTIAL UNITS  COMMERCIAL UNITS  TOTAL UNITS  \
0     10023                  0                 0            1   
1     10462                  1                 0            1   
2     11201                  1                 0            1   
3     11354                  1                 0            1   
4     10314                  1                 0            1   

   LAND SQUARE FEET  GROSS SQUARE FEET  YEAR BUILT  TAX CLASS AT TIME OF SALE  \
0       3941.675775        4045.707479           0                          2   
1          0.000000           0.000000           0                          1   
2          0.000000           0.000000           0                          1   
3       3941.675775        4045.707479        1920                          1   
4       3941.675775        4045.707479        1970                          1   

  BUILDING CLASS AT TIME OF SALE    SALE PRICE  SALE DATE  
0                             R4  1.276456e+06 2017-08-07  
1                             D4  0.000000e+00 2017-01-18  
2                             R4  0.000000e+00 2016-10-06  
3                             A1  1.276456e+06 2016-11-10  
4                             A5  1.276456e+06 2017-06-15

Group the DataFrame by the "BOROUGH" column and calculate the variance of "SALE PRICE" for each group

In [27]:
result = df_no_east.groupby('BOROUGH')['SALE PRICE'].var()
result

BOROUGH
1    4.587355e+14
2    7.745411e+12
3    1.549063e+13
4    7.274834e+12
5    2.694457e+12
Name: SALE PRICE, dtype: float64

Group the DataFrame by the "BOROUGH" column and calculate the standard deviation of "SALE PRICE" for each group

In [25]:
df_no_east.groupby('BOROUGH')['SALE PRICE'].std()

BOROUGH
1    2.141811e+07
2    2.783058e+06
3    3.935813e+06
4    2.697190e+06
5    1.641480e+06
Name: SALE PRICE, dtype: float64

Group the DataFrame by the "BOROUGH" column and perform aggregation on "SALE PRICE" for each group. Aggregation includes calculating mean, standard deviation, and the count of data points (size) for each borough


In [26]:
df_no_east.groupby('BOROUGH').agg({'SALE PRICE': [np.mean, np.std, np.size]})

SALE PRICE                     
                 mean           std   size
BOROUGH                                   
1        2.902477e+06  2.141811e+07  18306
2        5.901936e+05  2.783058e+06   7049
3        8.344884e+05  3.935813e+06  24047
4        9.063757e+05  2.697190e+06  26736
5        7.525600e+05  1.641480e+06   8410

In [28]:
mean_borough = 2.902477e+06
sd_borough = 2.141811e+07
n_borough = 18306
z = 1.96 # z-score untuk confidence level 95%

<b>The standard error (se) is calculated using the formula:</b>

se = sd / np.sqrt(jumlah keseluruhan data)

Where:
* se is the standard error.
* sd is the standard deviation.

Calculate with the standard error formula

In [30]:
se = sd_borough / np.sqrt(n_borough)

print("The result of the standard error is:", se)

The result of the standard error is: 158301.276054372


<b>Performing Confidence Interval Construction</b>

Formula:

lcb = mean_object - z * se  --> for the lower limit of the Confidence Interval (CI)

ucb = mean_object + z * se  --> for the upper limit of the Confidence Interval (CI)

<b>Explanation:</b>
- lcb: Lower limit of the Confidence Interval
- ucb: Upper limit of the Confidence Interval
- mean_object: Mean of the object (population or sample)
- z: Z-score for the desired confidence level
- se: Standard Error

Calculate the lower limit (lcb) and upper limit (ucb) for a confidence interval (CI).

In [31]:
lcb = mean_borough - z * se
ucb = mean_borough + z * se

print("Lower limit result:", lcb)
print("-------------------------")
print("Upper limit result:", ucb)

Lower limit result: 2592206.4989334308
-------------------------
Upper limit result: 3212747.5010665692


Grading Data

In [42]:
# Data 1
grade_borough_1 = df_no_east[df_no_east["BOROUGH"] == 1]

# Data 2
grade_borough_2 = df_no_east[df_no_east["BOROUGH"] == 3]

Calculating the number of data points

In [36]:
num_grade_borough_1 = len(grade_borough_1)
num_grade_borough_2 = len(grade_borough_2)

Calculating Mean

In [37]:
mean_grade_borough_1 = grade_borough_1["SALE PRICE"].mean()
mean_grade_borough_2 = grade_borough_2["SALE PRICE"].mean()

Calculating Standard Deviation

In [38]:
std_grade_borough_1 = grade_borough_1["SALE PRICE"].std()
std_grade_borough_2 = grade_borough_2["SALE PRICE"].std()

Display the results of calculations

In [39]:
print("Results for Data 1:", num_grade_borough_1, mean_grade_borough_1, std_grade_borough_1)
print("")
print("Results for Data 2:", num_grade_borough_2, mean_grade_borough_2, std_grade_borough_2)

Results for Data 1: 18306 2902476.5473534847 21418111.93479176

Results for Data 2: 24047 834488.3717719465 3935813.428569773


Summary

In [43]:

sm.stats.ztest(grade_borough_1["SALE PRICE"].dropna(), grade_borough_2["SALE PRICE"].dropna(), alternative="two-sided")

(14.651229205435005, 1.323003784566643e-48)

---

Overall Analysis:

From the analysis that has been done, information is obtained that the NYC property sales data has a lot of bad data so it is necessary to pre-process the data, the data tends to have a large distribution, and most of them show not normally distributed 